src/core/components/pin/gc-lock.js

```diff
@@ -0,0 +1,138 @@
+'use strict'
+
+const assert = require('assert')
+const mortice = require('mortice')
+const pull = require('pull-stream')
+const pullThrough = require('pull-stream/throughs/through')
+const pullAsyncMap = require('pull-stream/throughs/async-map')
+const EventEmitter = require('events')
+const log = require('debug')('ipfs:gc:lock')
+
+class GCLock extends EventEmitter {
```

Comment:
It simplifies testing, eg [whereever p-event is used](https://github.com/ipfs/js-ipfs/pull/2022/files#diff-c3c5bb981f284f576bb39b60a89014d4R76) in gc.spec.js

Reply:
please please don't use events, use signals or hooks instead 🙏 

http://blog.millermedeiros.com/callbacks-promises-signals-and-events/

this can be just a couple of hooks functions passed in the options, it will be much easier to debug, discover, mock, etc.. no need to clean listeners

Reply:
It would be nice to not have code in src that is exclusively for testing. Is there another way you can perform these tests?

Reply:
Fair enough - I will try writing it using sinon. Would that make sense to you Hugo?

Reply:
sinon is gonna be tricky for some of them but yeah it's better.

im trying to find a better solution but without a declarative way to handle the lifecycle of this code it's hard

Reply:
hmm yeah I'm not really seeing an easy way to do it, because the PullLocker object gets created when a pull lock is needed, but the lock is only taken when the pull stream starts

Reply:
Fresh out of Twitter https://twitter.com/davidmarkclem/status/1151540675093127168?s=20 check up and down the thread